plugins/permission-react/src/hooks/usePermission.ts

```diff
@@ -30,25 +33,40 @@ export type AsyncPermissionResult = {
 };
 
 /**
- * React hook utility for authorization. Given a
- * {@link @backstage/plugin-permission-common#Permission} and an optional
- * resourceRef, it will return whether or not access is allowed (for the given
- * resource, if resourceRef is provided). See
+ * React hook utility for authorization. Given either a non-resource
+ * {@link @backstage/plugin-permission-common#Permission} or a
+ * {@link @backstage/plugin-permission-common#ResourcePermission} and an
+ * optional resourceRef, it will return whether or not access is allowed (for
+ * the given resource, if resourceRef is provided). See
  * {@link @backstage/plugin-permission-common/PermissionClient#authorize} for
  * more details.
  *
+ * The resourceRef parameter is optional to allow calling this hook with an
+ * entity that might be loading asynchronously, but when resourceRef is not
+ * supplied, the value of `allowed` will always be false.
+ *
  * Note: This hook uses stale-while-revalidate to help avoid flicker in UI
  * elements that would be conditionally rendered based on the `allowed` result
  * of this hook.
  * @public
  */
-export const usePermission = (
-  permission: Permission,
-  resourceRef?: string,
-): AsyncPermissionResult => {
+export function usePermission(
+  ...[permission, resourceRef]:
+    | [ResourcePermission, string | undefined]
+    | [Exclude<Permission, ResourcePermission>]
+): AsyncPermissionResult {
```

Comment:
I messed around with this locally and the overloaded function signature seems to work with the current implementation. Thoughts on going down that path instead of making the arguments a tuple?

```ts
export function usePermission(
  permission: ResourcePermission,
  resourceRef: string | undefined,
): AsyncPermissionResult;
export function usePermission(
  permission: BasicPermission,
): AsyncPermissionResult;
export function usePermission(
  permission: Permission,
  resourceRef?: string | undefined,
) {
// current implementation
}
```

Reply:
This signature causes compilation errors in `PermissionedRoute` for me. I'm no expert on overloaded signatures, so very happy to keep looking for an alternative here - clearly the tuple signature is pretty gross :D.

Reply:
Ah yup I see that now. We'd have to overload `PermissionedRoute` as well to make that work. I'm starting to think the best way to reduce complexity here is to have a separate hook and route component for resource permissions...

Reply:
Yeah - we could do that! I wonder if it might make life a bit harder for consumers of the components, who today can seamlessly use the `PermissionedRoute` component for both the `<PermissionedRoute permission={catalogXPermission} resourceRef={currentEntity} />` and the `<PermissionedRoute permission={someOtherPermission} />` use case.

Reply:
Okay I took another look at this and tried refactoring the tuple to an object, which feels a bit easier to follow and matches the approach for `PermissionedRoute` props. Pushed it up a commit to demonstrate 3d8aacb8386b83f5ff53c0e052076e17ae0e718a

Reply:
Nice, this is definitely better! Will cherry-pick it in, thanks Joe! Might play around with adding support for passing a permission as well as an object - that way you can call `usePermission` via either `usePermission(permission)` or `usePermission({ permission, resourceRef })`.